File: Data Structures & Algorithms/count-number-of-islands/submission-3.java
class Solution {
    private final int[][] directions = {{1,0}, {-1,0}, {0,-1}, {0, 1}};
    public int numIslands(char[][] grid) {
        int row = grid.length;
        int col = grid[0].length;
        int result = 0;

        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (grid[i][j] == '1') {
                    dfs(i, j, grid);
                    result++;
                }
            }
        }
        return result;
    }

    private void dfs(int i, int j, char[][] grid) {
        int row = grid.length;
        int col = grid[0].length;
        if (i < 0 || i >= row || j < 0 || j >= col || grid[i][j] == '0') {
            return;
        }
        grid[i][j] = '0';
        for (int[] dir : directions) {
            dfs(i + dir[0], j + dir[1], grid);
        }
    }
}

File: Data Structures & Algorithms/max-area-of-island/submission-3.py
DIRECTIONS = [(-1, 0), (1, 0), (0, 1), (0, -1)]
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])

        def bfs(r, c):
            q = deque()
            grid[r][c] = 0
            q.append((r, c))
            res = 1

            while q:
                row, col = q.popleft()
                for dr, dc in DIRECTIONS:
                    nr, nc = row + dr, col + dc
                    if nr < 0 or nr >= m or nc < 0 or nc >= n or grid[nr][nc] == 0:
                        continue
                    res += 1
                    q.append((nr, nc))
                    grid[nr][nc] = 0
            return res

        area = 0    
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    area = max(area, bfs(i, j))
        return area            

            

                
